google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/ChangeStreamRecordAdapter.java

```diff
@@ -0,0 +1,172 @@
+
+  /**
+   * A SAX style change stream record factory. It is responsible for creating one of the three types
+   * of change stream record: heartbeat, close stream, and a change stream mutation.
+   *
+   * <p>State management is handled external to the implementation of this class:
+   *
+   * <ol>
+   *   Case 1: Heartbeat
+   *   <li>Exactly 1 {@code onHeartbeat}.
+   * </ol>
+   *
+   * <ol>
+   *   Case 2: CloseStream
+   *   <li>Exactly 1 {@code onCloseStream}.
+   * </ol>
+   *
+   * <ol>
+   *   Case 3: ChangeStreamMutation. A change stream mutation consists of one or more mods, where
+   *   the SetCells might be chunked. There are 3 different types of mods that a ReadChangeStream
+   *   response can have:
+   *   <li>DeleteFamily -> Exactly 1 {@code deleteFamily}
+   *   <li>DeleteCell -> Exactly 1 {@code deleteCell}
+   *   <li>SetCell -> Exactly 1 {@code startCell}, At least 1 {@code CellValue}, Exactly 1 {@code
+   *       finishCell}.
+   * </ol>
+   *
+   * <p>The whole flow of constructing a ChangeStreamMutation is:
+   *
+   * <ol>
+   *   <li>Exactly 1 {@code startUserMutation} or {@code startGcMutation}.
+   *   <li>At least 1 DeleteFamily/DeleteCell/SetCell mods.
+   *   <li>Exactly 1 {@code finishChangeStreamMutation}.
+   * </ol>
+   *
+   * <p>Note: For a non-chunked SetCell, only 1 {@code CellValue} will be called. For a chunked
+   * SetCell, more than 1 {@code CellValue}s will be called.
+   *
+   * <p>Note: DeleteRow's won't appear in data changes since they'll be converted to multiple
+   * DeleteFamily's.
+   */
+  interface ChangeStreamRecordBuilder<ChangeStreamRecordT> {
+    /**
+     * Called to create a heartbeat. This will be called at most once. If called, the current change
+     * stream record must not include any data changes or close stream messages.
+     */
+    ChangeStreamRecordT onHeartbeat(ReadChangeStreamResponse.Heartbeat heartbeat);
+
+    /**
+     * Called to create a close stream message. This will be called at most once. If called, the
+     * current change stream record must not include any data changes or heartbeats.
+     */
+    ChangeStreamRecordT onCloseStream(ReadChangeStreamResponse.CloseStream closeStream);
+
+    /**
+     * Called to start a new user initiated ChangeStreamMutation. This will be called at most once.
+     * If called, the current change stream record must not include any close stream message or
+     * heartbeat.
+     */
+    void startUserMutation(
+        @Nonnull ByteString rowKey,
+        @Nonnull String sourceClusterId,
+        @Nonnull long commitTimestamp,
+        int tieBreaker);
+
+    /**
+     * Called to start a new Garbage Collection ChangeStreamMutation. This will be called at most
+     * once. If called, the current change stream record must not include any close stream message
+     * or heartbeat.
+     */
+    void startGcMutation(@Nonnull ByteString rowKey, @Nonnull long commitTimestamp, int tieBreaker);
```

Comment:
here and everywhere else primitives can't be null

Reply:
Done for all places